File: common/array/java/ArrayPeer.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-5-31  Jed Young  Creation
//

package fan.fgfxArray;

import fan.sys.*;

public class ArrayPeer
{
  private Object array;
  private NumType type;

  public static ArrayPeer make(Array self) { return new ArrayPeer(); }

//////////////////////////////////////////////////////////////////////////
// fields
//////////////////////////////////////////////////////////////////////////

  public long size(Array self)
  {
    if (type == NumType.tByte)
    {
      return ((byte[])array).length;
    }
    else if (type == NumType.tFloat)
    {
      return ((float[])array).length;
    }
    else if (type == NumType.tDouble)
    {
      return ((double[])array).length;
    }
    else if(type == NumType.tInt)
    {
      return ((int[])array).length;
    }
    else if(type == NumType.tShort)
    {
      return ((short[])array).length;
    }
    else if(type == NumType.tLong)
    {
      return ((long[])array).length;
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }

  public NumType type(Array self) { return type; }

//////////////////////////////////////////////////////////////////////////
// ctor
//////////////////////////////////////////////////////////////////////////

  public static Array allocate(long size, NumType type)
  {
    Array array = Array.make();
    array.peer.type = type;

    int n = (int)size;
    if (type == NumType.tByte)
    {
      array.peer.array = new byte[n];
    }
    else if (type == NumType.tFloat)
    {
      array.peer.array = new float[n];
    }
    else if (type == NumType.tDouble)
    {
      array.peer.array = new double[n];
    }
    else if(type == NumType.tInt)
    {
      array.peer.array = new int[n];
    }
    else if(type == NumType.tShort)
    {
      array.peer.array = new short[n];
    }
    else if(type == NumType.tLong)
    {
      array.peer.array = new long[n];
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
    return array;
  }
  public static Array allocate(long size)
  {
    return allocate(size, NumType.tInt);
  }

//////////////////////////////////////////////////////////////////////////
// random read/write
//////////////////////////////////////////////////////////////////////////

  public long getInt(Array self, long index)
  {
    int i = (int)index;
    if (type == NumType.tByte)
    {
      return ((byte[])array)[i];
    }
    else if(type == NumType.tInt)
    {
      return ((int[])array)[i];
    }
    else if(type == NumType.tShort)
    {
      return ((short[])array)[i];
    }
    else if(type == NumType.tLong)
    {
      return ((long[])array)[i];
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }
  public Array setInt(Array self, long index, long v)
  {
    int i = (int)index;
    if (type == NumType.tByte)
    {
      ((byte[])array)[i] = (byte)v;
    }
    else if(type == NumType.tInt)
    {
      ((int[])array)[i] = (int)v;
    }
    else if(type == NumType.tShort)
    {
      ((short[])array)[i] = (short)v;
    }
    else if(type == NumType.tLong)
    {
      ((long[])array)[i] = (long)v;
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
    return self;
  }

  public double getFloat(Array self, long index)
  {
    int i = (int)index;
    if (type == NumType.tFloat)
    {
      return ((float[])array)[i];
    }
    else if (type == NumType.tDouble)
    {
      return ((double[])array)[i];
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }
  public Array setFloat(Array self, long index, double v)
  {
    int i = (int)index;
    if (type == NumType.tFloat)
    {
      ((float[])array)[i] = (float)v;
    }
    else if (type == NumType.tDouble)
    {
      ((double[])array)[i] = (double)v;
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
    return self;
  }

//////////////////////////////////////////////////////////////////////////
// batch read/write
//////////////////////////////////////////////////////////////////////////

  public static Array fromList(Array self, List list, NumType type)
  {
    Array array = Array.make();
    array.peer.array = ArrayUtil.listToArray(list, type);
    array.peer.type = type;
    return array;
  }

  public List toList(Array self)
  {
    if (type == NumType.tByte || type == NumType.tInt || type == NumType.tShort || type == NumType.tLong)
    {
      int size = (int)size(self);
      List list = List.make(Sys.IntType, size);
      for (int i = 0; i < size; ++i)
      {
        list.add(getInt(self, i));
      }
      return list;
    }
    else if (type == NumType.tFloat || type == NumType.tDouble)
    {
      int size = (int)size(self);
      List list = List.make(Sys.FloatType, size);
      for (int i = 0; i < size; ++i)
      {
        list.add(getFloat(self, i));
      }
      return list;
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }

//////////////////////////////////////////////////////////////////////////
// copy
//////////////////////////////////////////////////////////////////////////

  public void copyTo(Array self, Array dst, long dstOffset, long srcOffset, long size)
  {
    System.arraycopy(self.peer.array, (int)srcOffset, dst.peer.array, (int)dstOffset, (int)size);
  }
  public void copyTo(Array self, Array dst, long dstOffset, long srcOffset)
  {
    copyTo(self, dst, dstOffset, srcOffset, this.size(self));
  }
  public void copyTo(Array self, Array dst, long dstOffset)
  {
    copyTo(self, dst, dstOffset, 0);
  }
  public void copyTo(Array self, Array dst)
  {
    copyTo(self, dst, 0);
  }

//////////////////////////////////////////////////////////////////////////
// methods
//////////////////////////////////////////////////////////////////////////

  public Object getValue()
  {
    return array;
  }
}

File: common/array/js/ArrayPeer.js
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-5-29  Jed Young  Creation
//


fan.fgfxArray.ArrayPeer = fan.sys.Obj.$extend(fan.sys.Obj);
fan.fgfxArray.ArrayPeer.prototype.$ctor = function(self) {}
fan.fgfxArray.ArrayPeer.prototype.array = null;

//////////////////////////////////////////////////////////////////////////
// native
//////////////////////////////////////////////////////////////////////////

fan.fgfxArray.ArrayPeer.prototype.size = function(self) { return this.array.length; }

fan.fgfxArray.ArrayPeer.prototype.type = function(self) { return this.m_type; };

//////////////////////////////////////////////////////////////////////////
// ctor
//////////////////////////////////////////////////////////////////////////

fan.fgfxArray.ArrayPeer.allocate = function(size, type)
{
  var a = fan.fgfxArray.Array.make();
  if (!type) type = fan.fgfxArray.NumType.m_tInt;
  a.peer.m_type = type;
  a.peer.array = new Array(size);
  return a;
}

//////////////////////////////////////////////////////////////////////////
// random read/write
//////////////////////////////////////////////////////////////////////////

fan.fgfxArray.ArrayPeer.prototype.getInt = function(self, index)
{
  if (this.type(self) != fan.fgfxArray.NumType.m_tInt) throw UnsupportedErr.make("not Int buffer");
  return this.array[index];
}
fan.fgfxArray.ArrayPeer.prototype.setInt = function(self, index, v)
{
  if (this.type(self) != fan.fgfxArray.NumType.m_tInt) throw UnsupportedErr.make("not Int buffer");
  this.array[index] = v;
  return self;
}
fan.fgfxArray.ArrayPeer.prototype.getFloat = function(self, index)
{
  if (this.type(self) != fan.fgfxArray.NumType.m_tFloat) throw UnsupportedErr.make("not Float buffer");
  return this.array[index];
}
fan.fgfxArray.ArrayPeer.prototype.setFloat = function(self, index, v)
{
  if (this.type(self) != fan.fgfxArray.NumType.m_tFloat) throw UnsupportedErr.make("not Float buffer");
  this.array[index] = v;
  return self;
}

//////////////////////////////////////////////////////////////////////////
// batch read/write
//////////////////////////////////////////////////////////////////////////

fan.fgfxArray.ArrayPeer.prototype.fromList = function(self, list)
{
  var size = list.size();
  var array
  if (list.of == fan.sys.Int.$type)
  {
    array = fan.fgfxArray.Array.allocate(size, fan.fgfxArray.NumType.m_tInt);
  }
  else if(list.of == fan.sys.Float.$type)
  {
    array = fan.fgfxArray.Array.allocate(size, fan.fgfxArray.NumType.m_tFloat);
  }
  else
  {
    throw UnsupportedErr.make("Unknow type: " + list.of);
  }

  for (var i = 0; i < size; ++i)
  {
    array.peer.array[0] = list.get(0);
  }
  return array;
}
fan.fgfxArray.ArrayPeer.prototype.toList = function(self)
{
  var size = this.size(self);
  var type = this.type(self);

  if (type == NumType.tByte || type == NumType.tInt || type == NumType.tShort || type == NumType.tLong)
  {
    var list = fan.sys.List.make(fan.sys.Int.$type, size);
    for (var i = 0; i < size; ++i)
    {
      list.add(this.getInt(self, i));
    }
    return list;
  }
  else if (type == NumType.tFloat || type == NumType.tDouble)
  {
    var list = fan.sys.List.make(fan.sys.Float.$type, size);
    for (var i = 0; i < size; ++i)
    {
      list.add(this.getFloat(self, i));
    }
    return list;
  }
  else
  {
    throw UnsupportedErr.make("unsupported type");
  }
}
fan.fgfxArray.ArrayPeer.prototype.copyTo = function(self, dst, dstOffset, srcOffset, size)
{
  if (!dstOffset) dstOffset = 0;
  if (!srcOffset) srcOffset = 0;
  if (!size) size = this.size(self);

  if (self.type() != dst.type()) throw ArgErr.make("Must have same type");
  if (size > self.size()-srcOffset || size > dst.size()-dstOffset) throw IndexErr.make("out of range: "+size);

  for(var i=0; i < size; ++i)
  {
    dst.peer.array[i+dstOffset] = this.array[i+srcOffset];
  }
}

//////////////////////////////////////////////////////////////////////////
// methods
//////////////////////////////////////////////////////////////////////////

fan.fgfxArray.ArrayPeer.prototype.getValue = function()
{
  return this.array;
}